src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt

```diff
@@ -236,12 +240,33 @@ private constructor(private val project: Project, newSessionId: CompletableFutur
     @RequiresEdt
     fun createNew(project: Project): AgentChatSession {
       val sessionId = createNewPanel(project) { it.server.chatNew() }
-      val chatSession = AgentChatSession(project, sessionId)
+      val chatSession = AgentChatSession(project, UUID.randomUUID().toString(), sessionId)
+        println("chat session added internalId=${chatSession.internalId}")
       synchronized(chatSessions) { chatSessions.add(chatSession) }
       return chatSession
     }
 
-    private fun createNewPanel(
+      fun createFromState(project: Project, state: ChatState): AgentChatSession {
+          val sessionId = createNewPanel(project) { it.server.chatNew() }
+          val chatSession = AgentChatSession(project, state.internalId!!, sessionId)
+          chatSessions.add(chatSession)
```

Comment:
Please synchronise this and add it as a last thing in this method. As quickly as you add it, it is possible to receive messages into this session, and we usually od not want that until all else is initialised.

Reply:
Good point! Fixed.